### PR TITLE
Stop voice capture when the selected session closes

### LINKED ISSUE #2131: [Detail Bug] Voice chat: mic keeps recording after session is closed, causing silent loss of transcripts
# Detail Bug Report

https://app.detail.dev/org_89974092-6629-40ef-8b02-01754f4ef520/bugs/bug_b99c7d72-5f9c-4615-acc1-6af611aa38b5

Introduced in [#1656](https://github.com/purplefish-ai/factory-factory/pull/1656) by @martin-purplefish on May 29, 2026

# Summary
- **Context**: `VoiceModeToggle` is a React component that enables voice-based chat by capturing microphone input and sending transcripts to the current session.
- **Bug**: When the session is deleted/closed while voice capture is active, the microphone continues capturing and transcripts are silently discarded. The rejection-recovery mechanism CANNOT recover messages with null sessionId.
- **Actual vs. expected**: The mic should stop capturing when the session is deleted, but instead it continues running. Transcripts are sent, rejected, and immediately cleared without recovery.
- **Impact**: User's spoken input is permanently lost without any error feedback or recovery option.

# Code with Bug
```tsx
// src/client/features/voice/voice-mode-toggle.tsx

// Effect 1: Only stops capture on unmount, not on sessionId change
useEffect(() => {
  return () => stop();
}, [stop]);  // <-- BUG 🔴 missing sessionId dependency

// The button stays clickable when capturing (intentional), but there's no
// auto-stop when sessionId becomes null
const handleClick = useCallback(() => {
  if (isCapturing || isConnecting) {
    stop();
    setVoiceModeOn(false);
  } else {
    primeAudioContext();
    setVoiceModeOn(true);
    void start();  // <-- BUG 🔴 starts capture without checking sessionId
  }
}, [isCapturing, isConnecting, start, stop, primeAudioContext]);
```

```ts
// src/client/features/chat/use-chat-state.ts

useEffect(() => {
  if (state.lastRejectedMessage) {
    const { text, attachments, sessionId } = state.lastRejectedMessage;
    if (!sessionId || sessionId !== dbSessionIdRef.current) {
      dispatch({ type: 'CLEAR_REJECTED_MESSAGE' });  // <-- BUG 🔴 null sessionId rejected messages are cleared without recovery
      return;
    }
    // Recovery logic (never reached for null sessionId)
    setInputDraft(text);
    if (attachments && attachments.length > 0) {
      setInputAttachments(attachments);
    }
    dispatch({ type: 'CLEAR_REJECTED_MESSAGE' });
  }
}, [state.lastRejectedMessage, setInputDraft, setInputAttachments]);
```

# Explanation
- Closing a session sets the selected sessionId to `null`, but the chat UI (and `VoiceModeToggle`) stays mounted (it is hidden, not unmounted), so the “stop on unmount” cleanup never runs.
- While capture is still active, `onFinalTranscript` continues to call `sendMessage(...)` with `sessionId: null`. The backend rejects these messages.
- The rejected-message recovery path cannot recover `sessionId: null` messages because `use-chat-state.ts` immediately clears `lastRejectedMessage` when `!sessionId`, so transcripts are discarded with no user-visible feedback.

## Codebase Inconsistency
`VoiceModeToggle` already implements an auto-stop pattern for invalid states (config disables voice mode):
```ts
useEffect(() => {
  if (config !== undefined && !voiceModeAvailable && (isCapturing || isConnecting)) {
    stop();
    setVoiceModeOn(false);
  }
}, [config, voiceModeAvailable, isCapturing, isConnecting, stop]);
```
Session invalidation (`sessionId` becoming `null`) is another invalid state but is not handled similarly.

# Recommended Fix
Add an effect in `VoiceModeToggle` to stop capture when `sessionId` becomes `null` while capturing/connecting:
```ts
useEffect(() => {
  if (!sessionId && (isCapturing || isConnecting)) {
    stop();
    setVoiceModeOn(false);
  }
}, [sessionId, isCapturing, isConnecting, stop]);
```

# History
This bug was introduced in commit ce58491d. The commit attempted to harden rejected message recovery by preventing recovery when there's no valid session ID (`!sessionId` check added at line 212). This was intended as a security measure to avoid recovering messages without a session context. However, it inadvertently broke voice mode recovery: when a user deletes the last session while voice capture is in progress, sessionId becomes null, the recovery effect fires, and the `!sessionId` guard skips recovery entirely—permanently losing the user's voice transcript. The fix added a test case for the intended behavior but missed the voice mode edge case where an in-flight capture's session is deleted.

### PR DESCRIPTION
**Pull request by martin-purplefish**
Closing or switching the selected chat session now stops active microphone capture or an in-progress connection attempt and disables voice playback. Speech cannot continue into another selected session while the component stays mounted. The closed-session story uses a local tRPC transport so it does not depend on a running backend.

Closes #2131.

Validation: `pnpm check:fix`, `pnpm typecheck`, `pnpm test src/client/features/voice/voice-mode-toggle.test.tsx src/client/features/voice/use-mic-capture.test.ts` (19 tests), `pnpm check`, `pnpm check:imports`, and `pnpm build:storybook` passed. Closure and non-null switch regressions cover both capturing and connecting. Headless Chromium verified the built story on initial load and reload: the Voice Off toggle is disabled, with zero backend fetch/XHR requests and zero page errors. Pre-commit checks passed. The local Codex schema check skips CLI 0.154.0 versus pinned 0.153.4; CI enforces it.

### CODE REVIEW

File: src/client/features/voice/voice-mode-toggle.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+const meta: Meta<typeof VoiceModeToggle> = {
+  title: 'Voice/VoiceModeToggle',
+  component: VoiceModeToggle,
+  args: { onFinalTranscript: fn() },
+  decorators: [
+    (Story) => {
+      const utils = trpc.useUtils();
+      useEffect(() => {
+        utils.voice.getConfig.setData(undefined, {
```
[the remark this answers is not in the document]
Comment (martin-purplefish) [reply src/client/features/voice/voice-mode-toggle.stories.tsx]: Fixed in c9e414b1. The decorator now supplies its own tRPC client with a local observable transport and QueryClientProvider, so no backend request can be issued for configuration. Storybook build and typecheck pass. Headless Chromium verified the built closed-session story on initial load and reload: the Voice Off button is disabled, with zero backend fetch/XHR requests and zero page errors.

File: src/client/features/voice/voice-mode-toggle.tsx [hunk 2]
```diff
@@ -154,6 +154,13 @@ export function VoiceModeToggle({
     return () => stop();
   }, [stop]);
 
+  useEffect(() => {
+    if (!sessionId && (isCapturing || isConnecting)) {
+      stop();
+      setVoiceModeOn(false);
+    }
```
[the remark this answers is not in the document]
Comment (martin-purplefish) [reply src/client/features/voice/voice-mode-toggle.tsx:164]: Fixed in c9e414b1. Capture now stops on a change to another non-null session as well as closure, while rerendering the same session leaves capture alone. Added capturing and connecting regressions for both transitions; the new non-null switch cases failed before the fix. All 19 focused voice tests plus check:fix, typecheck, check, check:imports, and pre-commit checks pass.
